Primary system indices should be write indices for system aliases

williamrandolph opened the pull request:
Previously, when auto-creating a system index, we would assign the alias from the system index descriptor to any system index that matched the descriptor's pattern. This meant that, as a side effect, a reindex operation or unexpected autocreate operation can add an index to an alias and cause it to reject writes.

I could think of two fixes. The first was to adjust our code so that a system alias is only automatically added to the system index descriptor's primary index. ~That is what I have done in this pull request.~

The other option is to ensure that the primary index is always designated as the system alias's write index. **EDIT: That is what I have done in this pull request.**

- [x] Will the system index manager propagate this update to existing aliases?

Fixes https://github.com/elastic/elasticsearch/issues/85072

williamrandolph committed:
Do not autocreate alias for non-primary system indices

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

williamrandolph committed:
Update docs/changelog/85977.yaml

elasticsearchmachine commented:
Hi @williamrandolph, I've created a changelog YAML for you.

williamrandolph committed:
Merge branch 'master' into system-indices-only-autocreate-alias-for-primary

rjernst commented:
> The other option would be to find a way to ensure that the primary index is always designated as the system alias's write index

For non-externally managed system indices, this seems like the more correct solution? The intent of the system indices infrastructure is to completely manage system indices, so it would seem aligned to have the infrastructure completely control the alias.

williamrandolph committed:
Merge branch 'master' into system-indices-only-autocreate-alias-for-primary

williamrandolph committed:
Primary sysindex is always write index

I mistakenly believed that the only way to set an index as write index
for an alias was to go through the Alias API. I found out I was wrong
about that, which makes it significantly easier to change the approach
on this branch. Now, whenever we create a system index whose descriptor
has an alias, we add the alias. If the system index is the primary
index, we make it the write index for the alias.

williamrandolph commented:
All of the production code changes here are on code paths that only apply to managed system indices, so on second thought it is probably all right to be restrictive.

I've changed the code so that we always add the alias for any managed system index, and set the primary index as the write index for the alias.

I need to look at the system index manager code to see if we expect these changes to propagate on a version upgrade.

williamrandolph commented:
The `SystemIndexManager` class can handle mapping updates, but it doesn't do anything to perform fairly trivial fixes to system indices when the system index descriptors change.

We should probably file a different issue to cover this work. It could involve checking that the index is hidden, that its alias is hidden, and that its primary index is designated as the write index for its alias. One reason I'd want to spin this work off is that there could be some complexity around making sure that nodes with different versions aren't competing with each other to change index settings.

grcevski commented on server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/AutoCreateSystemIndexIT.java:
```diff
@@ -84,6 +99,24 @@ public void testAutoCreateNonPrimaryIndex() throws Exception {
+        assertThat(response.aliases().get(INDEX_NAME + "-2").get(0), equalTo(AliasMetadata.builder(INDEX_NAME).isHidden(true).build()));
+    }
+
+    public void testAutoCreateSystemIndicesAllowsWriteToAlias() throws Exception {
```
Would it be possible to test the opposite scenario? Such that, the first API request makes the non primary index INDEX_NAME-2 and then we add the primary name - 1.

grcevski commented on server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:
```diff
@@ -98,9 +100,29 @@ public void testSystemIndexIsAutoCreatedViaConcreteName() {
+        assertFalse(indexExists(PRIMARY_INDEX_NAME));
+        assertTrue(indexExists(INDEX_NAME + "-2"));
+
+        // Check that a non-primary system index is not automatically assigned an alias
```
I'm a little bit confused by the comment here, I guess there's an alias, but it's not an alias we can use for writing?

williamrandolph replied:
Good catch. That comment is out of date. I'll change it!

grcevski reviewed (approved):
LGTM! This looks good to me, I just left some minor comments and questions?

williamrandolph committed:
Merge branch 'master' into system-indices-only-autocreate-alias-for-primary

williamrandolph committed:
Test different ordering for index creation

williamrandolph committed:
Fix code comment

williamrandolph commented:
@elasticmachine please run elasticsearch-ci/part-1

williamrandolph commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into system-indices-only-autocreate-alias-for-primary

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.2](https://github.com/elastic/elasticsearch/pull/86769)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/86769">](https://github.com/elastic/elasticsearch/pull/86769) |
